### DISCUSSION TITLE
`tl.loader()` performance implications?

### DISCUSSION THREAD

**Discussion by mrjones2014** [teal-language/tl#513] [Q&A] [answered]
Hi, I'm very interested in Teal, it looks awesome. I'm considering rewriting parts of my Neovim plugin in Teal to help with documentation (I think there's some docs generators for Teal), and I'm trying to plan my approach.

I see the main approach used by Neovim plugins written in Teal is to commit the compiled Lua to source control (because of the way Neovim plugin managers work, the plugin files need to be right in the repo). This is fine, but I'd like to avoid the need to store compiled files in source control, which brings me to my question about `tl.loader()` and possibly using that instead.

What are the performance implications of `tl.loader()`? Can you speak a little to how it works? Are there benchmarks?

**Comment by euclidianAce** [comment, marked as answer]
`tl.loader()` injects a 'searcher' function as described [here in the Lua manual](https://www.lua.org/manual/5.4/manual.html#pdf-package.searchers) to search for `.tl` files when using `require` and compiles them on the fly (_without_ type checking them). As far as performance goes, I personally wouldn't recommend it for Neovim as it can increase your start up time quite a bit (obviously depending on how much code you are compiling).

But as far as not wanting to commit the generated Lua, you have a couple options:
 - many Neovim plugin managers provide a post-install hook that you could use to compile the plugin. This of course requires the user to have `tl` or `cyan` installed (which isn't a problem for packer, not sure if other managers have luarocks deps yet).
 - you could use github actions (or something) to build and zip up the lua files to distribute your compiled plugin

Personally I don't mind just committing the generated Lua, you could even set up a [pre commit hook](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks) to compile before committing.

Anyway, you can find (incomplete) type definitions for Neovim's Lua api [here in the teal-types repo](https://github.com/teal-language/teal-types) if you do decide to use Teal.

**Reply by mrjones2014** [reply]
Thanks for that, yeah I definitely don't want to use the loader then. I'll set up a git hook.

**Comment by danielo515** [comment]
What about using it for personal nvim configs? Is anyone using teal for that so I can take inspiration from them? I'm a bit tired of all the guess work while coding in Lua for neovim
